File: payload/reggae/reggae.rb
require 'json'

# Find the build object
module BuildFinder
  def self.get_build
    builds = []
    ObjectSpace.each_object(Build) { |x| builds << x }
    if builds.length != 1
      fail "Only one Build object may exist, found #{builds.length}"
    end
    builds[0]
  end
end

# Aggregates top-level targets
class Build
  def initialize(*targets)
    @targets = targets
  end

  def to_json
    jsonify.to_json
  end

  def jsonify
    @targets.map { |t| t.jsonify }
  end
end

def build(*targets)
  Build.new(*targets)
end

# A build target
class Target
  attr_reader :outputs, :command, :dependencies, :implicits

  def initialize(outputs, command = '', dependencies = [], implicits = [])
    @outputs = arrayify(outputs)
    @command = jsonifiable(command, ShellCommand)
    @dependencies = dependify(dependencies, FixedDependencies)
    @implicits = dependify(implicits, FixedDependencies)
  end

  def to_json
    jsonify.to_json
  end

  def jsonify
    { type: 'fixed',
      command: @command.jsonify,
      outputs: @outputs,
      dependencies: @dependencies.jsonify,
      implicits: @implicits.jsonify
    }
  end
end

def target(outputs, command = '', dependencies = [], implicits = [])
  Target.new(outputs, command, dependencies, implicits)
end

# A shell command
class ShellCommand
  def initialize(cmd = '')
    @cmd = cmd
  end

  def jsonify
    @cmd == '' ? {} : { type: 'shell', cmd: @cmd }
  end
end

private def arrayify(arg)
  arg.class == Array ? arg : [arg]
end

private def jsonifiable(arg, klass)
  (arg.respond_to? :jsonify) ? arg : klass.new(arg)
end

private def dependify(arg, klass)
  if arg.is_a? Dependencies
    return arg
  end

  if arg.is_a?(Array) && arg.length > 1 && \
     arg.any? { |x| x.is_a? DynamicDependencies }
    return target_concat(arg)
  end

  klass.new(arg)
end

private def target_concat(targets)
  DynamicDependencies.new('targetConcat', dependencies: targets.map { |x| x.jsonify })
end

# Equivalent to link in the D version
class LinkCommand
  def initialize(flags)
    @flags = flags
  end

  def jsonify
    { type: 'link', flags: @flags }
  end
end

def link(exe_name:, flags: '', dependencies: [], implicits: [])
  Target.new([exe_name], LinkCommand.new(flags), dependencies, implicits)
end

def object_files(src_dirs: [], exclude_dirs: [],
                 src_files: [], exclude_files: [],
                 flags: '',
                 includes: [], string_imports: [])
  DynamicDependencies.new('objectFiles',
                          { src_dirs: src_dirs,
                            exclude_dirs: exclude_dirs,
                            src_files: src_files,
                            exclude_files: exclude_files,
                            flags: flags,
                            includes: includes,
                            string_imports: string_imports })
end

class Dependencies
end

# A 'compile-time' known list of dependencies
class FixedDependencies < Dependencies
  def initialize(deps)
    @deps = arrayify(deps)
  end

  def jsonify
    { type: 'fixed', targets: @deps.map { |t| t.jsonify } }
  end
end

# a rule to create a static library
def static_library(name,
                   src_dirs: [],
                   exclude_dirs: [],
                   src_files: [],
                   exclude_files: [],
                   flags: '',
                   includes: [],
                   string_imports: [])
    DynamicDependencies.new('staticLibrary',
                            { name: name,
                              src_dirs: src_dirs,
                              exclude_dirs: exclude_dirs,
                              src_files: src_files,
                              exclude_files: exclude_files,
                              flags: flags,
                              includes: includes,
                              string_imports: string_imports })
end

def scriptlike(src_name:,
               exe_name:,
               flags: '',
               includes: [],
               string_imports: [],
               link_with: [])

  Dynamic.new('scriptlike',
              { src_name: src_name,
                exe_name: exe_name,
                flags: flags,
                includes: includes,
                string_imports: string_imports,
                link_with: dependify(link_with, FixedDependencies) })
end

# A run-time determined list of dependencies
class DynamicDependencies < Dependencies
  def initialize(func_name, args)
    @func_name = func_name
    @args = args
  end

  def jsonify
    base = { type: 'dynamic', func: @func_name }
    base.merge(@args)
  end
end

# dynamic target
class Dynamic
  def initialize(func_name, args)
    @func_name = func_name
    @args = args
  end

  def jsonify
    hash = { type: 'dynamic', func: @func_name }
    @args.each do |k, v|
      if v.respond_to? :jsonify
        hash[k] = v.jsonify
      else
        hash[k] = v
      end
    end
    hash
  end
end

File: payload/reggae/reggae.lua
local JSON = require('JSON')

local Build = {}
Build.__index = Build
setmetatable(Build, {
                __call = function(cls, ...)
                   return cls.new(...)
                end,
})

local Target = {}
Target.__index = Target

setmetatable(Target, {
                __call = function (cls, ...)
                   return cls.new(...)
                end,
})

local ShellCommand = {}
ShellCommand.__index = ShellCommand

setmetatable(ShellCommand, {
                __call = function (cls, ...)
                   return cls.new(...)
                end
})



local LinkCommand = {}
LinkCommand.__index = LinkCommand

setmetatable(LinkCommand, {
                __call = function (cls, ...)
                   return cls.new(...)
                end
})


local FixedDependencies = {}
FixedDependencies.__index = FixedDependencies

setmetatable(FixedDependencies, {
                __call = function (cls, ...)
                   return cls.new(...)
                end
})

local DynamicDependencies = {}
DynamicDependencies.__index = DynamicDependencies

setmetatable(DynamicDependencies, {
                __call = function (cls, ...)
                   return cls.new(...)
                end
})


function Build.new(target)
   local self = setmetatable({}, Build)
   self.targets = {target}
   self.isBuild = true
   return self
end

function Build:to_json()
   return JSON:encode(self:jsonify())
end

function Build:jsonify()
   targets = {}
   for k, v in pairs(self.targets) do
      targets[k] = v:jsonify()
   end

   return targets
end

function Target.new(outputs, cmd, deps, imps)
   local self = setmetatable({}, Target)

   cmd = cmd or ''

   self.command = jsonifiable(cmd, ShellCommand)
   self.outputs = arrayify(outputs)
   self.dependencies = dependify(deps, FixedDependencies)
   self.implicits = dependify(imps, FixedDependencies)

   return self
end

function Target:to_json()
   return JSON:encode(self:jsonify())
end

function Target:jsonify()
   return {
           type = 'fixed',
           command = self.command:jsonify(),
           outputs = self.outputs,
           dependencies = self.dependencies:jsonify(),
           implicits = self.implicits:jsonify(),
   }
end

function jsonifiable(arg, cls)
   return (arg and arg.jsonify) and arg or cls.new(arg)
end

function dependify(arg, cls)
   return (arg and arg.isDependency) and arg or cls.new(arg)
end

function arrayify(arg)
   if arg == nil then
      return {}
   end

   if isArray(arg) then
      return arg
   else
      return {arg}
   end
end

function isArray(arg)
   if type(arg) ~= 'table' then
      return false
   end

   for k, v in pairs(arg) do
      if type(k) ~= 'number' then
         return false
      end
   end

   return true
end

function ShellCommand.new(cmd)
   local self = setmetatable({}, ShellCommand)
   self.cmd = (cmd == '') and {} or {type='shell', cmd=cmd}
   return self
end

function ShellCommand:jsonify()
   return self.cmd
end

function FixedDependencies.new(deps)
   local self = setmetatable({}, FixedDependencies)
   self.isDependency = true
   self.targets = arrayify(deps)
   return self
end


function FixedDependencies:jsonify()
   local targets = {}
   for k, v in pairs(self.targets) do
      targets[k] = v:jsonify()
   end
   return {type = 'fixed', targets = targets}
end

function link(options)
   options.flags = options.flags or ''
   options.dependencies = options.dependencies or {}
   options.implicits = options.implicits or {}

    return Target.new(options.exe_name,
                      LinkCommand.new(options.flags),
                      options.dependencies,
                      options.implicits)

end

function LinkCommand.new(flags)
   local self = setmetatable({}, LinkCommand)
   self.flags = flags
   return self
end

function LinkCommand:jsonify()
   return {type = 'link', flags = self.flags}
end

function object_files(options)

   options.src_dirs = options.src_dirs or {}
   options.exclude_dirs = options.exclude_dirs or {}
   options.src_files = options.src_files or {}
   options.exclude_files = options.exclude_files or {}
   options.flags = options.flags or ""
   options.includes = options.includes or {}
   options.string_imports = options.string_imports or {}

   return DynamicDependencies.new('objectFiles', options)
end

function DynamicDependencies.new(func_name, args)
   local self = setmetatable({}, DynamicDependencies)

   self.isDependency = true
   self.func_name = func_name
   self.args = args

   return self
end

function DynamicDependencies:jsonify()
   local result = {type = 'dynamic', func = self.func_name}

   for k, v in pairs(self.args) do
      if v.jsonify then
         result[k] = v:jsonify()
      else
         result[k] = v
      end
   end

   return result
end

function static_library(name, options)
   options.name = name
   options.src_dirs = options.src_dirs or {}
   options.exclude_dirs = options.exclude_dirs or {}
   options.src_files = options.src_files or {}
   options.exclude_files = options.exclude_files or {}
   options.flags = options.flags or ""
   options.includes = options.includes or {}
   options.string_imports = options.string_imports or {}

   return DynamicDependencies('staticLibrary', options)
end

function scriptlike(options)

    options.flags = options.flags or ""
    options.includes = options.includes or {}
    options.string_imports = options.string_imports or {}
    options.link_with = options.link_with or FixedDependencies.new({})

    return DynamicDependencies.new('scriptlike', options)
end


return {
   Build = Build,
   Target = Target,
   link = link,
   object_files = object_files,
   static_library = static_library,
   scriptlike = scriptlike,
}
